libgadget/timebinmgr.c

```diff
@@ -95,27 +96,60 @@ void set_sync_params(int OutputListLength, double * OutputListTimes)
  * integer stamps.
  **/
 void
-setup_sync_points(double TimeIC, double TimeMax, double no_snapshot_until_time, int SnapshotWithFOF)
+setup_sync_points(double TimeIC, double TimeMax, double UVBGTimestep, double no_snapshot_until_time, int SnapshotWithFOF)
 {
     int i;
 
     qsort_openmp(Sync.OutputListTimes, Sync.OutputListLength, sizeof(double), cmp_double);
 
     if(NSyncPoints > 0)
         myfree(SyncPoints);
-    SyncPoints = mymalloc("SyncPoints", sizeof(SyncPoint) * (Sync.OutputListLength+2));
+    //TODO(Jdavies): don't use syncpoints for uvbg calculation, or figure out how many are there beforehand
+    //z=20 to z=4 is ~150 syncpoints at 10 Myr spaces
+    //
+    SyncPoints = mymalloc("SyncPoints", sizeof(SyncPoint) * (Sync.OutputListLength+2+400)); 
 
     /* Set up first and last entry to SyncPoints; TODO we can insert many more! */
-
+    //NOTE(jdavies): these first syncpoints need to be in order
+   
     SyncPoints[0].a = TimeIC;
     SyncPoints[0].loga = log(TimeIC);
     SyncPoints[0].write_snapshot = 0; /* by default no output here. */
     SyncPoints[0].write_fof = 0;
-    SyncPoints[1].a = TimeMax;
-    SyncPoints[1].loga = log(TimeMax);
-    SyncPoints[1].write_snapshot = 1;
-    SyncPoints[1].write_fof = 0;
-    NSyncPoints = 2;
+    SyncPoints[0].calc_uvbg = 0;
+    NSyncPoints = 1;
+
+    // UVBG calculation every 10 Myr from z=20
+    {
+        double z_start = 20.;
+        double uv_a = 1.0 / (1.0 + z_start);
+        while (uv_a <= TimeMax) {
+            SyncPoints[NSyncPoints].a = uv_a;
+            SyncPoints[NSyncPoints].loga = log(uv_a);
+            SyncPoints[NSyncPoints].write_snapshot = 0;
+            SyncPoints[NSyncPoints].write_fof = 0;
+            SyncPoints[NSyncPoints].calc_uvbg = 1;
+            NSyncPoints++;
+            //message(0,"added UVBG syncpoint at a = %.3f, Nsync = %d\n",uv_a,NSyncPoints);
+
+            // TODO(smutch): OK - this is ridiculous (sorry!), but I just wanted to quickly hack something...
+            // TODO(jdavies): fix low-z where delta_a > 10Myr
+            double delta_a = 0.0001;
+            double lbt = time_to_present(uv_a);
+            double delta_lbt = 0.0;
```

Comment:
The minimum time between UVBG calculations is set by the UVBGTimestep, this loop just steps through scale factors until we hit a point where that much time has past, and sets the next syncpoint there to avoid more integration.

Reply:
Yup - I mean: what is the physical reason that you set UVBGTimestep to the value that it has? Why is 10Myr the important value?

Reply:
10Myr was originally chosen because it was a small enough cadence to cover the lifespan of the OB stars that produced most of the UV photons

Reply:
Aha, I see, thank you. So if we find that this is an expensive part of the simulation, is there still the option to increase UVBGTimestep? Or have you found that doing that causes the results to not be physically valid anymore?

Reply:
Actually I might have gotten a bit confused, that timestep was chosen for the lifetime of the large stars, but I think it was mostly relevant for the supernovae feedback from those stars in Meraxes. I haven't done much testing of the timestep, and because we plug in the rates from the excursion set into the existing cooling model, the way the timestep affects things might change. I suspect it will be okay as long as the density or position of gas, or the size of the bubbles doesn't change drastically within the timestep.

Reply:
Right - one reason I'm asking is that the PMstep has a displacement constraint: a new PM step is triggered whenever particles move through 0.1 of a grid cell. The PM grid cells are much smaller than the UVBG cells, so it might be enough to use the PM timestep (which I think is longer than 10Myr). But I'm not sure about the bubble size, which I guess increases at the speed of light?

Reply:
We could definitely try increasing the timestep to see if that works, ionisation fronts move at speeds anywhere between 1e2 km/s in dense regions and 1e5 km/s in the voids during overlap. If we have a reionisation scenario that mostly happens between z=8 to z=6 we have about 300Myr, which would be ~30 steps at a 10Myr spacing, meaning optimistically every step covers at most  ~3-4% of reionisation. I don't think we'd want to go much coarser than this, otherwise particles might stay neutral for a significant portion of the EoR. However, I'd need to test longer steps before understanding how it would effect the evolution of the particles.

Reply:
That makes sense, thanks for the information!